### PR TITLE
Refine QS for manually connecting Node.js

### PR DESCRIPTION
**Pull request by jbyrne-redhat**

### CODE REVIEW

File: nodejs/README.adoc [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+====
+
+// Condition out QS-only content so that it doesn't appear in docs.
+// All QS anchor IDs must be in this alternate anchor ID format `[#anchor-id]` because the ascii splitter relies on the other format `[id="anchor-id"]` to generate module files.
+ifdef::qs[]
+[#description]
+Learn how to use Node.js applications to produce and consume messages using a Kafka instance in {product-long}.
+
+[#introduction]
+Welcome to the quick start for {product-long} with Node.js. In this quick start, you'll learn how to use https://nodejs.org/en/about/[Node.js^] to produce messages to and consume messages from your Kafka instances in {product}.
```
Comment (wtrocki) [inline nodejs/README.adoc:69]: use Node.js sounds strange here. I would rephrase it to something like: "learn to use Node.js runtime to..

File: nodejs/README.adoc [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+
+[id="proc-importing-nodejs-sample-code_{context}"]
+== Importing the Node.js sample code
+
+For this quick start, you'll use sample code from the Nodeshift Application Starters https://github.com/nodeshift-starters/reactive-example[reactive-example^] repository in GitHub. After you understand the concepts and tasks in this quick start, you can use your own Node.js applications with {product} in the same way.
+
+.Procedure
+. On your computer, open a command-line window.
+. On the command line, clone the Nodeshift Application Starters https://github.com/nodeshift-starters/reactive-example[reactive-example^] repository from GitHub.
```
Comment (wtrocki) [inline nodejs/README.adoc:80]:
```suggestion
. On the command line, clone the Nodeshift Application Starter https://github.com/nodeshift-starters/reactive-example[reactive-example^] repository from GitHub.
```

File: nodejs/README.adoc [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+.Prerequisites
+* You've created a Kafka instance in {product} and the instance is in the *Ready* state.
+
+.Procedure
+. In the {product} web console, go to *Streams for Apache Kafka* > *Kafka Instances* and click the name of the Kafka instance that you want to add a topic to.
+. Click *Create topic* and follow the guided steps to define the topic details. Click *Next* to complete each step and click *Finish* to complete the setup.
++
+[.screencapture]
+.Guided steps to define topic details
```
Comment (wtrocki) [inline nodejs/README.adoc:160]: This repeats a lot with other guides. Worth checking if content is the same

File: nodejs/README.adoc [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+.Procedure
+. On the command line, navigate to the `reactive-example` directory corresponding to the repository that you cloned.
++
+.Navigating to the reactive-example directory
+[source]
+----
+$ cd ~/reactive-example
+----
+
+. Check out the `kafkajs` branch of the repository.
```
Comment (wtrocki) [inline nodejs/README.adoc:201]: I think this is the default branch right now when we checkout source code is it might not be needed.

WDYT @mhdawson ?

Ref:
https://github.com/nodeshift-starters/reactive-example

File: nodejs/README.adoc [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+----
+$ source set-environment.sh
+----
+
+. Navigate to the directory for the producer component. Use Node Package Manager (npm) to install the package for this component.
++
+.Installing the producer component
+[source]
+----
+$ cd producer-backend
```
Comment (wtrocki) [inline nodejs/README.adoc:267]: While those steps are easy it felt that we could have them grouped together with some script on the repository.
For example npm workspace can run multiple projects at the same time with the proper logs.

File: nodejs/README.adoc [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+----
+$ git checkout kafkajs
+----
+
+. Run the shell script that sets environment variables for the bootstrap server endpoint and client credentials.
++
+.Running the shell script
+[source]
+----
+$ source set-environment.sh
```
Comment (wtrocki) [inline nodejs/README.adoc:259]: This is antipattern for node.js.

Would it be possible to use something like https://www.npmjs.com/package/dotenv in the repo and remove those steps from guide?

**Comment by wtrocki** [comment]
Done end to end check (I had some problems with rendering so used plain ascidocs view).
Changes look amazing. Left some possible improvements

File: nodejs/README.adoc (hunk 4 above)
Comment (mhdawson) [reply nodejs/README.adoc:201]: I had just changed that recently I think @jbyrne-redhat was planning to remove it.

File: nodejs/README.adoc (hunk 6 above)
Comment (mhdawson) [reply nodejs/README.adoc:259]: I did that to avoid having people having to export the variables multiple times while still going over what you need to do so that people would understand and can apply to whatever they will do in their environment.  I looks like using a .env file might be similar.  They key thing for me is that the person going through it knows what they need to set/use in their code.

File: nodejs/README.adoc (hunk 3 above)
Comment (mhdawson) [reply nodejs/README.adoc:160]: The version in the doc I created was copied from the Quarkus guide, except for the topic create.

File: nodejs/README.adoc (hunk 5 above)
Comment (mhdawson) [reply nodejs/README.adoc:267]: I sometimes find that when its just run this script then that script you don't see what is going on. Since these are quite simple I"d lean towards leaving the individual steps. They don't take long and you see the exact steps.

File: nodejs/README.adoc (hunk 6 above)
Comment (mhdawson) [reply nodejs/README.adoc:259]: The nice thing about using .env would be that it should work the same across platforms so for that reason I think it's worth doing.

I'll see if I can find tomorrow to adjust for using .env.

Comment (wtrocki) [reply nodejs/README.adoc:259]: Thank you. .Env is used by Quarkus guide as well.

File: nodejs/README.adoc (hunk 5 above)
Comment (wtrocki) [reply nodejs/README.adoc:267]: Agreed.

File: nodejs/README.adoc (hunk 3 above)
Comment (jbyrne-redhat) [reply nodejs/README.adoc:160]: Yes, we are using the same content as for other quick starts (for example, Quarkus), but with some minor tweaks.

File: nodejs/README.adoc (hunk 4 above)
Comment (jbyrne-redhat) [reply nodejs/README.adoc:201]: Yes, given that `kafkajs` is now the default branch, I've removed the step that explicitly instructs users to check it out.

File: nodejs/README.adoc (hunk 6 above)
Comment (jbyrne-redhat) [reply nodejs/README.adoc:259]: @mhdawson  Let me know if you implement this change. Right now, we instruct users to create a shell script for Linux/macOS, or a batch file for Windows.

File: nodejs/README.adoc (hunk 2 above)
Comment (wtrocki) [reply nodejs/README.adoc:80]: No strong opinion. I just found it odd when reading.
That name reflects organization. What we cloning is repository which is instance of single starter.

Comment (jbyrne-redhat) [reply nodejs/README.adoc:80]: Yes, sorry, _organization_ was the word I was looking for - not project. Regardless, I think we are right to name it as it appears.

File: nodejs/README.adoc (hunk 6 above)
Comment (mhdawson) [reply nodejs/README.adoc:259]: @jbyrne-redhat will do, pretty busy but might find time on Friday , will let you know.

Comment (mhdawson) [reply nodejs/README.adoc:259]: From https://access.redhat.com/documentation/en-us/red_hat_openshift_streams_for_apache_kafka/1/guide/04827d87-ed92-4ffd-a126-11fa13348eba looks to me like the quarkus guide uses environment variables, but still think its a good idea to look at the .env approach for the Node.js guide.

File: nodejs/README.adoc [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+[id="proc-configuring-nodejs_{context}"]
+== Configuring the Node.js example application to connect to a Kafka instance
+
+To enable your Node.js application to access a Kafka instance, you need to configure a connection by specifying the following details:
+
+* The bootstrap server endpoint for your Kafka instance
+* The generated credentials for your {product} service account
+* The Simple Authentication and Security Layer (SASL) mechanism that the client will use to authenticate with the Kafka instance
+
+In this task, you'll create a new shell script called `set-environment.sh`. In this shell script, you'll set the required bootstrap server and client credentials as environment variables.
```
Comment (mhdawson) [inline nodejs/README.adoc:103]:
```
In this task, you'll create a new file called `.env`. In this file you'll set the required bootstrap server and client credentials so that they can be used to connect to your Kafka instance.
```

File: nodejs/README.adoc [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+.Prerequisites
+ifndef::qs[]
+* You have the bootstrap server endpoint and client credentials for the Kafka instance. You can access this information as follows:
+** For the bootstrap server endpoint, select the options menu (three vertical dots) for a Kafka instance in {product}. Click *View connection information*.
+** For the client ID and client secret, you're prompted to copy and store these credentials when you create a service account in {product}.
+endif::[]
+
+.Procedure
+
+. In your IDE, open a new file for the shell script. Save the file with the name `set-environment.sh`.
```
Comment (mhdawson) [inline nodejs/README.adoc:114]:
```suggestion
. In your IDE, open a new file for the shell script. Save the file with the name `.env`.
```

File: nodejs/README.adoc [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+
+. In the `set-environment.sh` file, add the lines shown in the example to set the bootstrap server and client credentials as environment variables to be used by the Node.js application.
```
Comment (mhdawson) [inline nodejs/README.adoc:116]:
```suggestion
. In the `.env` file, add the lines shown in the example to set the bootstrap server and client credentials as environment variables to be used by the Node.js application.
```

File: nodejs/README.adoc [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
++
+.Setting environment variables for bootstrap server and client credentials
+[source,subs="+quotes"]
+----
+export KAFKA_BOOTSTRAP_SERVER=__<bootstrap_server>__
+export KAFKA_CLIENT_ID=__<client_id>__
+export KAFKA_CLIENT_SECRET=__<client_secret>__
+export KAFKA_SASL_MECHANISM=plain
```
Comment (mhdawson) [inline nodejs/README.adoc:124]:
```suggestion
KAFKA_BOOTSTRAP_SERVER=__<bootstrap_server>__
KAFKA_CLIENT_ID=__<client_id>__
KAFKA_CLIENT_SECRET=__<client_secret>__
KAFKA_SASL_MECHANISM=plain
```

File: nodejs/README.adoc [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+--
+<bootstrap_server>:: The bootstrap server endpoint for your service account. To access this information for a Kafka instance in {product}, select the options menu (three vertical dots). Click *View connection information*.
+<client_id>:: A client credential generated when you create a service account in {product}. You're prompted to copy and store this credential when you create the service account.
+<client_secret>:: A client credential generated when you create a service account in {product}. You're prompted to copy and store this credential when you create the service account.
+--
+endif::[]
++
+In this case, observe that the Node.js application uses the `SASL/PLAIN` authentication method (that is, the value of `KAFKA_SASL_MECHANISM` is set to `plain`). This means that the application uses only the client ID and client secret to authenticate with the Kafka instance. The application *doesn't* also require an authentication token.
+
+. Save the the `set-environment.sh` file.
```
Comment (mhdawson) [inline nodejs/README.adoc:140]:
```suggestion
. Save the the `.env` file.
```

File: nodejs/README.adoc [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+----
+
+. Run the shell script that sets environment variables for the bootstrap server endpoint and client credentials. These are required to connect to your Kafka instance.
++
+.Running the shell script
+[source]
+----
+$ source set-environment.sh
+----
+
```
Comment (mhdawson) [inline nodejs/README.adoc:216]:
```suggestion
```

File: nodejs/README.adoc (hunk 6 above)
Comment (mhdawson) [inline nodejs/README.adoc:259]:
```suggestion
```

File: nodejs/README.adoc [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+$ git checkout kafkajs
+----
+
+. Run the shell script that sets environment variables for the bootstrap server endpoint and client credentials.
++
+.Running the shell script
+[source]
+----
+$ source set-environment.sh
+----
```
Comment (mhdawson) [inline nodejs/README.adoc:260]:
```suggestion
```

File: nodejs/README.adoc [hunk 14]
```diff
@@ -0,0 +1,341 @@
...
+Bolivia
+Haiti
+----
++
+The output from both components shows that the Node.js application successfully connected to the Kafka instance. The application is using the Kafka topic you created to produce and consume messages.
+
+. To verify __how__ the Node.js application connected to the Kafka instance, perform the following actions:
+.. In your IDE, open the `producer.js` file.
++
+Observe that the producer component is using the environment variables that you set in the `set-environment.sh` shell script. The values of these environment variables are the bootstrap server endpoint and client credentials that the producer component used to connect to the Kafka instance.
```
Comment (mhdawson) [inline nodejs/README.adoc:322]:
```suggestion
Observe that the producer component is using the environment variables that you set in the `.env` file. The values of these environment variables are the bootstrap server endpoint and client credentials that the producer component used to connect to the Kafka instance.

```

Comment (jbyrne-redhat) [reply nodejs/README.adoc:322]: Thanks a lot for making these updates, @mhdawson. I made some other updates on my branch in the meantime, so I'll incorporate your updates also. I should be able to push all those changes sometime Monday and will ask you for a final review then.

File: nodejs/README.adoc (hunk 6 above)
Comment (mhdawson) [reply nodejs/README.adoc:259]: PR to update the kafka version to support dotenv: https://github.com/nodeshift-starters/reactive-example/pull/34. Working on node-rdkafka version.

Comment (mhdawson) [reply nodejs/README.adoc:259]: PR for node-rdkafka - https://github.com/nodeshift-starters/reactive-example/pull/35

Both PRs are landed, and with the suggestions I made to swtich to creating '.env' file I think I'm done.

@jbyrne-redhat let me know if you need anything else.

**Comment by jbyrne-redhat** [comment]
@mhdawson I've applied your updates and finished my own refinement. It all ran nicely for me. Can you give this a final once over?

**Comment by wtrocki** [comment]
@jbyrne-redhat do you want another pair of eyes for final ack?

**Review by wtrocki — APPROVED** [review]

**Comment by jbyrne-redhat** [comment]
> @jbyrne-redhat do you want another pair of eyes for final ack?

Thanks a lot, @wtrocki!

**Comment by wtrocki** [comment]
This looks and feels really well. Thank you @jbyrne-redhat @mhdawson for amazing work!

**Review by mhdawson — APPROVED** [review]
LGTM

**Comment by mhdawson** [comment]
@jbyrne-redhat added my LGTM looks good, thanks for all your work on this.

**Comment by jbyrne-redhat** [comment]
@b1zzu Per https://issues.redhat.com/browse/MGDSTRM-4779, this quick start is ready for final QE verification.  An HTML preview is available at http://file.emea.redhat.com/~jbyrne/j223/README.html. Thanks a lot!

**Comment by wtrocki** [comment]
Verified

**Review by b1zzu — APPROVED** [review]
Verified and tested locally

**Comment by wtrocki** [comment]
Good to merge
